src/puppetlabs/puppetdb/queue.clj

```diff
@@ -0,0 +1,199 @@
+  (stock/discard q (:entry command)))
+
+(deftype SortedCommandBuffer [^TreeMap fifo-queue ^HashMap certnames-map ^long max-entries ^clojure.lang.IFn delete-update-fn]
+  async-protos/Buffer
+  (full? [this]
+    (>= (.size fifo-queue) max-entries))
+
+  (remove! [this]
+    (let [^CommandRef cmdref (val (.pollFirstEntry fifo-queue))
+          command-type (:command cmdref)]
+      (when (or (= command-type "replace catalog")
+                (= command-type "replace facts"))
+        (.remove certnames-map [command-type (:certname cmdref)]))
+      cmdref))
+
+  (add!* [this item]
+    (when-not (instance? CommandRef item)
+      (throw (IllegalArgumentException. (str "Cannot enqueue item of type " (class item)))))
+
+    (let [^CommandRef cmdref item
+          command-type (:command cmdref)
+          certname (:certname cmdref)]
+
+      (when (or (= command-type "replace catalog")
+                (= command-type "replace facts"))
+        (when-let [^CommandRef old-command (.get certnames-map [command-type certname])]
+          (.put fifo-queue
+                (:id old-command)
+                (assoc old-command :delete? true))
+          (delete-update-fn (:command old-command) (:version old-command)))
+        (.put certnames-map [command-type certname] cmdref))
+      (.put fifo-queue (:id cmdref) cmdref))
+    this)
+
+  (close-buf! [this])
+  clojure.lang.Counted
+  (count [this]
+    (.size fifo-queue)))
+
+(defn sorted-command-buffer
+  ([^long n]
+   (sorted-command-buffer n (constantly nil)))
+  ([^long n ^clojure.lang.IFn delete-update-fn]
+   ;; accepting a function here is a hack to get around a cyclic dependency
+   ;; between this ns, mq-listener.clj, and dlo.clj. My hope is we'll be able
+   ;; to get rid of it somehow when we refactor mq-listener and command.clj.
+   (SortedCommandBuffer. (TreeMap.) (HashMap.) n delete-update-fn)))
+
+(defn message-loader
+  "Returns a function that will enqueue existing stockpile messages to
+  `command-chan`. Messages with ids less than `message-id-ceiling`
+  will be loaded to guard against duplicate processing of commands
+  when new commands are enqueued before all existing commands have
+  been enqueued. Note that there is no guarantee on the enqueuing
+  order of commands read from stockpile's reduce function"
+  [q message-id-ceiling]
+  (fn [command-chan update-metrics]
+    (stock/reduce q
+                  (fn [chan entry]
+                    ;;This conditional guards against a new command
+                    ;;enqueued in the same directory before we've full
+                    ;;read all existing files
+                    (when (< (stock/entry-id entry) message-id-ceiling)
+                      (let [{:keys [command version] :as cmdref} (entry->cmdref entry)]
+                        (async/>!! chan cmdref)
+                        (update-metrics command version)))
+                    chan)
+                  command-chan)))
+
+(defn create-or-open-stockpile
```

Comment:
There's two pieces to opening a stockpile instance. The opening of the queue itself so that you can enqueue more stuff, then there's what to do with the existing messages. This returns the Stockpile queue instance and a function that can be called when ready to read in the existing messages from the queue. This allows enqueuing of old messages will starting up (and/or received new messages).